Add configuration for only posting warnings to added/modified lines

Phillita opened the pull request:
First off, thank you for your work on Pronto/Pronto Rubocop!

Problem:

I'm looking to add functionality to this gem that hides the external Rubocop warnings for unchanged lines of code. 
This is to help control the intake of warnings that are generated by new code without having to resolve warnings around it.

Example:

The file in question has a "Class has too many lines" Rubocop warning. 
Updating a method within this class that didn't cause this warning directly, is now receiving a Github Review comment with this warning.

Solution:

- Add configuration for "only_patched_lines"
    - `enabled: true || false`
    - `range: n` where n is added to the warning start to broaden how many warnings are caught
- If a warning has been generated by a new addition, it will be included in the review comments
- If the warning is outside of the new code being added, ignore it for this review
- If a warning range has been provided, capture warnings that start within that range from the patched lines.

Caveats:

If new code generates a warning that starts outside of the patched lines, and no range has been provided, it will be ignored.
The range option has been added to capture some of these but not all will be caught.

Getting a delta of warnings added before the new code would be the best outcome here, but that seemed non-trivial.


UPDATE - April 18, 2024

We've been running into issues with confusing offences showing up on lines that they are not directly connected to. To handle this, I've added in some messaging to show the line that the message/suggestion applies to. I've also made it so that suggestions will only be given to lines where the offence directly maps to the code line.

UPDATE - June 18, 2024

I've added in documentation links from the github pull request.

Phillita committed:
Add configuration for only posting warnings to added/modified lines

Phillita committed:
Allow warnings to be caught within a range

Phillita committed:
Update README.md

Co-authored-by: Bert McCutchen <bert.mccutchen@gmail.com>

Phillita commented on README.md:
```diff
@@ -39,10 +51,34 @@ For example:
+  def too_long
+    # Pretend I am a new method that's 32 lines long.
+  end
+end
+
```
```suggestion
end
```

```

Phillita committed:
End code example

Phillita committed:
Add messaging for indirect offenses

Phillita committed:
Exctract suggestion into new class to help with ruby critic.

Phillita committed:
Remove some linter warnings. Add support folder to rspec

Phillita committed:
Add a documentation url to messages

ramil350 committed:
fix deprecated method warnings with Rubocop 1.65+

ashkulz committed:
update README and gemspec

ashkulz committed:
add Ruby 3.3 and 3.4 to the CI matrix

ashkulz committed:
setup Dependabot and Trusted Publishing workflow
see https://guides.rubygems.org/trusted-publishing/

ashkulz committed:
bump version to 0.11.6

ashkulz committed:
push_gem: fix repository name

ashkulz commented:
@Phillita can you rebase your PR on the latest `master`?

Phillita committed:
Merge branch 'master' into comments-for-changed-lines-only

Phillita commented:
> @Phillita can you rebase your PR on the latest `master`?

@ashkulz I tried to rebase master onto my branch but I obviously did it wrong and reverted back to a merge. It's up to date with the new changes.

ashkulz commented:
@Phillita I'm not sure the merge was correct, because it hasn't run the checks -- I normally get an option to approve them, but on this PR even that option is missing.

Phillita commented:
> > @Phillita I'm not sure the merge was correct, because it hasn't run the checks -- I normally get an option to approve them, but on this PR even that option is missing.
> 
> Strange! It says that the workflow requires approval from a maintainer. I'm not quite sure what that means but I'll try looking into it when I get to my laptop.

@ashkulz Does it allow you to trigger the workflow run? I'm seeing this in the [Github documentation](https://docs.github.com/en/actions/managing-workflow-runs-and-deployments/managing-workflow-runs/approving-workflow-runs-from-public-forks#approving-workflow-runs-on-a-pull-request-from-a-public-fork) but I have limited knowledge on this:
> If you are comfortable with running workflows on the pull request branch, return to the  Conversation tab, and under "Workflow(s) awaiting approval", click Approve and run.

ashkulz commented:
It doesn't, this is what I see:

![Screenshot 2025-01-13 at 21-31-50 Add configuration for only posting warnings to added_modified lines by Phillita · Pull Request #89 · prontolabs_pronto-rubocop](https://github.com/user-attachments/assets/07e5839e-417a-4d66-88a5-7f4d51fa045a)

ashkulz commented:
Ah, there was an extra button in the ["GitHub Actions" run](https://github.com/prontolabs/pronto-rubocop/actions/runs/12726957757) and now it seems to be running:

![Screenshot 2025-01-13 at 21-34-11 Add configuration for only posting warnings to added_modified lines · prontolabs_pronto-rubocop@1fdafa8](https://github.com/user-attachments/assets/efad113d-ce02-4d73-998f-139a5c73b4a3)

ashkulz reviewed (changes requested):
I'd appreciate if you could fix the failing tests before I did a thorough review on the PR, since that could cause further changes.

Phillita committed:
Remove byebug

Phillita committed:
Remove rubocop from the gemspec.

Rubocop has been defined in the gemfile for development and test use.

Phillita commented:
> I'd appreciate if you could fix the failing tests before I did a thorough review on the PR, since that could cause further changes.

I pushed a fix for the duplicate gem. I missed this being added during the merge. Locally the tests run.
It looks like each run will need to be approved before it executes @ashkulz

Phillita commented:
I understand what's going on now. The workflow adds specific versions of rubocop to the gemfile. This conflicts with me moving the gems out of the gemspec for local dev.

I'm going to try pushing up a change to the workflow to control this through bundler. I'm not sure if this will work but it's worth a shot.

Phillita committed:
Reverted rubocop in the gemfile and moved it back to the gemspec

Phillita commented:
@ashkulz nevermind, I reverted my gemfile change for rubocop. Let's see if the actions are happy with this.

Phillita committed:
Add base64 to the gemfile

Phillita commented:
Base64 has been added back into the gemfile. 1 last run will hopefully do the trick

ashkulz commented on Gemfile:
```diff
@@ -6,3 +6,10 @@ gemspec
 
 local_gemfile = File.expand_path('Gemfile.local', __dir__)
 eval_gemfile local_gemfile if File.exist?(local_gemfile)
+
+group :development, :test do
+  gem 'base64', '~> 0.1.2'
+  gem 'rake', '~> 12.0'
+  gem 'rspec', '~> 3.4'
+  gem 'rspec-its', '~> 1.3'
+end
```
**question**: what's the rationale to move it out of the gemspec into the `Gemfile`? All the other pronto gems do it via the gemspec, and not sure what benefits that brings.

bert-mccutchen replied:
I can help explain this change. 

Gemspec should only include dependencies needed at runtime for the gem. By moving these development dependencies to the Gemfile we can reduce unnecessary dependency conflicts with projects that implement the gem.

bert-mccutchen replied:
Ooh sorry, I was viewing this on my phone. And didn't notice it was originally a development_dependency instead of a runtime_dependency! There should really be no difference that I know of then.

Phillita replied:
For this one, I had been following a rubocop warning. I'm not seeing it in my editor anymore though
https://docs.rubocop.org/rubocop/cops_gemspec.html#gemspecdevelopmentdependencies

ashkulz replied:
It's not enabled by default, and it also supports the current convention via `EnforcedStyle: gemspec` so I think it's one of those opinionated ones it's safe to ignore.

ashkulz commented on lib/pronto/rubocop/offense_line.rb:
```diff
@@ -3,6 +3,8 @@
 module Pronto
   class Rubocop < Runner
     class OffenseLine
+      DOCS = 'https://docs.rubocop.org/rubocop/cops_%s.html#%s'
+
```
**concern**: I don't want to use the "default" URL, I'd rather we use `RuboCop::Version.document_version` to go to the version-specific documentation URL. One hiccup is that right now, [1.70.0 is the current version](https://github.com/rubocop/rubocop/releases/tag/v1.70.0) (_released 4 days ago_) but that's not reflected on the documentation site. Also, 0.85 is the earliest version for which they have documentation.

Phillita replied:
I would like your thoughts further on this comment. 

I agree pointing to the specific documentation version would be amazing but, without having the certainty that it will be there for the version of Rubocop being used, it could hinder rather than help.

I dug into this a bit. I figured there was probably a better way to get this link.
The following can generate the URL without the need to point to a constant:

```ruby
def documentation_url
  patch_cop.registry.find_by_cop_name(offense.cop_name)&.documentation_url
end
```

This still uses the main URL but leaves the logic with Rubocop. Thoughts?

ashkulz replied:
@Phillita apparently they do the docs build manually, rubocop/docs.rubocop.org#24 proposes to automate it so I think we _can_ assume that it'll be there.

I think it's neat that they have the `#documentation_url` method, but how far back does it go in terms of supported API?

Phillita replied:
It looks like the documentation_url goes back as far as rubocop v0.90.0.

ashkulz replied:
I'm happy to bump up the minimum version to when this was introduced, it'll also simplify a lot of the conditional logic we have and potentially not need to split it off to another class.

ashkulz commented on lib/pronto/rubocop/offense_line.rb:
```diff
@@ -26,81 +28,39 @@ def processed_source
-
-        @suggestion_text ||= corrected_lines[offense.line - 1]
+      def offense_message
+        offense.message.gsub(
+          offense.cop_name, "[#{offense.cop_name}](#{documentation_url})"
+        )
```
**concern**: does this affect CLI output as well? I'd want to differentiate between what is shown on the CLI vs what's posted as Markdown.

Phillita replied:
This will affect the CLI. Do you have a suggestion for how we can differentiate the output format based on where pronto is being run?

Here is an example run on some staged test code locally:
![image](https://github.com/user-attachments/assets/a9361562-0765-45f5-872b-96ebf37d6e84)

I don't think this is a bad thing, though. It provides an easy way to get to the documentation from my terminal.

ashkulz replied:
In the CLI, I'm more interested in seeing compact output so that it doesn't go over 80 (_or whatever terminal width people have configured_). Maybe we can keep `message_text` as-is (_after removing `suggestion_text`_) and use it for the CLI and introduce `message_markdown` (_which includes `suggestion_text` and these changes_) and use it in the appropriate formatters?

I know that's a large refactoring, but I think it'd be valuable in this case -- since most of the time suggestions are not enabled in the CLI but only in CI, this change would have an impact on people who use pronto via pre-commit hooks.

Phillita replied:
I'll have to look into how pronto passes formatters to the runners. I'm unsure about this at the moment. 
I agree it would be much more compact. I have a lot of experience with Github but I lack knowledge with bitbucket/gitlab. I don't know which ones would use markdown. 

It makes me wonder, would it be a good idea to add a config flag that turns markdown on and let the end user to control this?

ashkulz replied:
I think all of them use Markdown, if you do it for GitHub I think it'll be obvious how to do it for the others as well. I don't think we need a configuration flag as such, it should depend on the formatter.

Phillita replied:
@ashkulz I think I need a little more direction here. (Sorry in advance if I'm being dense)

When [pronto runs](https://github.com/prontolabs/pronto/blob/master/lib/pronto.rb), each runner is called without formatter context. 

Looking at how the github formatter works, it uses the messages from the runners without any additional parsing.
It creates [Comment objects here](https://github.com/prontolabs/pronto/blob/master/lib/pronto/formatter/git_formatter.rb#L71).

I'm confused about how to use the chosen formatter to get the proper message from the runner (with/without markdown).

Do you have context for how I could do this?

ashkulz commented on lib/pronto/rubocop/offense_line.rb:
```diff
@@ -128,6 +88,10 @@ def config_severities
         error: :error,
         fatal: :fatal
       }.freeze
+      SUGGESTION = 'suggestion'
+      RUBY = 'ruby'
+      INDIRECT_MESSAGE = "Offense generated for line %d:\n\n"
+      INDIRECT_SUGGESTION = "Suggestion for line %d:\n\n"
```
**suggestion**: please remove the constants which aren't used. In fact, I'd prefer if we included the messages inline as not seeing in any point in having it separate.

Phillita replied:
I usually write code where most strings are placed into constants for maintainability and code documentation.
If you'd prefer, I can move the strings inline. 

I'll get rid of the unused ones too.

ashkulz replied:
In this case, the constant name and the method name are the same so I'd prefer if it was inline (_so that part feels like a nit-picky, I'm fine either way_) -- but please remove the unused ones.

ashkulz commented on lib/pronto/rubocop/offense_suggestion.rb:
```diff
@@ -0,0 +1,117 @@
+
+module Pronto
+  class Rubocop < Runner
+    class OffenseSuggestion
```
**question**: what's the rationale on introducing this class?

Phillita replied:
I found this class kept the offense_line class much smaller and cleaner. The suggestion logic made sense to break out into its own self-contained class.

ashkulz replied:
This makes the diff larger and a bit harder to review. I'm neutral on this, so if you feel strongly I'll review it in detail a bit later.

Phillita replied:
I think this class makes sense to keep. If I move it back into the offense_line class, I think we'll hit linter warnings for the class being too long. 
That being said, if you do end up feeling strongly about this class and don't mind the other class becoming long, I can make the change.

ashkulz replied:
See the other suggestion -- a lot of the conditional logic might simply fall away, so I'd like you to re-evaluate your opinion. I'm not too concerned about the class being too long if it's all in one place.

Phillita replied:
I'll start moving the code back into the offense_line class

Phillita committed:
Revert Gemfile/Gemspec gem changes

Phillita committed:
Move the string contants inline and remove unused.

Phillita committed:
Use the built in documentation_url from rubocop

Bump the minimum version of rubocop for the gem to the earliest date for the documentation_url method.

Phillita committed:
Merge offense_suggestion into offense_line

Updated specs. Removed conditional code based on min rubocop version update.

Phillita committed:
Remove the require offense_suggestion

Phillita committed:
Update checks to remove anything below rubocop 0.90.0

Phillita commented:
@ashkulz Sorry for the delay. I've removed checks for anything below rubocop 0.90.0 since we've updated the base version for the gem.

Phillita commented:
> @Phillita sorry, but due to a family medical emergency I'm not going to be able to review this immediately. I'll try by this weekend :crossed_fingers:

No problem at all. I hope everything is ok.

ashkulz commented on .github/workflows/checks.yml:
```diff
@@ -39,9 +34,6 @@ jobs:
     - uses: actions/checkout@v4
     - name: use specific rubocop version
       run: echo "gem 'rubocop', '${{ matrix.rubocop }}'" > Gemfile.local
-    - name: use specific rubocop-ast version (if required)
-      if: matrix.rubocop == '0.84.0' || matrix.rubocop == '0.85.0'
-      run: echo "gem 'rubocop-ast', '< 0.7.0'" >> Gemfile.local
```
**thought**: since we're dropping support for these rubocop versions, might as well remove the section from the README as it's no longer relevant.

ashkulz commented on .github/workflows/checks.yml:
```diff
@@ -15,12 +15,7 @@ jobs:
       matrix:
         ruby: ['2.4', '2.5', '2.6', '2.7', '3.0', '3.1', '3.2', '3.3', '3.4']
-        rubocop: ['0.63.1', '0.81.0', '0.84.0', '0.85.0', '0.86.0', '0.87.0', '< 1.0', '1.0.0', '1.29.1', '1.30.0', '1.38.0', '< 2.0']
-        include:
-          - ruby: '2.3'
-            rubocop: '0.63.1'
-          - ruby: '2.3'
-            rubocop: '0.81.0'
```
**suggestion**: can you try and see if `0.90.0` works with Ruby 2.3 via the `include`? Otherwise, I'd recommend we bump up the `required_ruby_version` in the gemspec as well. It is a breaking change in a way, but I think Ruby 2.3 support (which was EOL ~6 years ago) isn't so important for us.

Phillita replied:
Let me try adding it to limit the breaking change by a bit.

ashkulz commented on pronto-rubocop.gemspec:
```diff
@@ -16,25 +16,22 @@ Gem::Specification.new do |s|
   s.add_development_dependency('rake', '~> 12.0')
   s.add_development_dependency('rspec', '~> 3.4')
-  s.add_development_dependency('rspec-its', '~> 1.2')
+  s.add_development_dependency('rspec-its', '~> 1.3')
```
**question**: any particular reason for bumping the minimum version up?

Phillita replied:
Hmm I think this was just me updating things as I normally would. I can revert it

ashkulz commented on pronto-rubocop.gemspec:
```diff
@@ -16,25 +16,22 @@ Gem::Specification.new do |s|
     spec/.*
     |Gemfile
     |Rakefile
     |\.rspec
     |\.gitignore
     |\.rubocop.yml
     |\.travis.yml
-    )$}x
-  end
-  s.test_files = []
+    )$}x)
```
**question**: what's the rationale for this change? Let's please revert these changes if it was just fixing warnings, as I don't want to touch the `gemspec` much at all. Happy to do it in a separate PR.

Phillita replied:
This was linter warning fixes. They can be reverted

ashkulz commented on lib/pronto/rubocop/offense_line.rb:
```diff
@@ -26,83 +26,100 @@ def processed_source
+        report.offenses.map(&:line).uniq.size
       end
 
+      # rubocop >= 0.87.0 has mobilize as a public method
```
**suggestion**: let's remove this and the above comment about `0.87.0` since they make no sense now.

ashkulz reviewed (changes requested):
Did a quick review, will also reply to [your other comment](https://github.com/prontolabs/pronto-rubocop/pull/89#discussion_r1919123951) later on in the week.

Phillita committed:
Remove gemspec rubocop changes

Phillita committed:
Remove readme section for old rubocop versions

Phillita committed:
Add ruby 2.3 back to the CI checks for the min rubocop version

Phillita committed:
Remove unneeded comments from the offense line

Phillita commented:
@ashkulz I've applied your suggestions. Review whenever you are able. No rush!

ashkulz commented on .github/workflows/checks.yml:
```diff
@@ -15,12 +15,10 @@ jobs:
-            rubocop: '0.63.1'
-          - ruby: '2.3'
-            rubocop: '0.81.0'
+            rubocop: '0.90.0'
```
**suggestion**: since this is failing CI, let's remove Ruby 2.3 from the matrix and bump up `required_ruby_version` in the `gemspec`. Also, I think rather than using `exclude` we should `include` specifically supported RuboCop versions for Ruby 2.4 and Ruby 2.5 and have Ruby 2.6+ in the matrix. What do you think?

Phillita replied:
I'm a fan of bumping up the version. 

For the include, it does make it easier to figure out which checks are running for that version of ruby. You don't have to parse as much to understand it. I'll push a change now.

Phillita committed:
Bump up the required ruby version to 2.4

Phillita committed:
CI checks updated for ruby 2.3 removal

Remove ruby 2.3 from the matrix. Include ruby 2.4/2.5 checks instead of using the exclude rules.

Phillita commented:
@ashkulz It's been a while! I hope everything is all good on your end. Please let me know if you'd like to continue discussion.

ashkulz commented:
@Phillita sorry for the delay, things have been crazy. I'll get back to this over the weekend.

Phillita commented:
@ashkulz sorry but my workplace has switched away from pronto. This is no longer actively in use.
The documentation url part is close to working but I think it fails in a few spots. (It generates some broken urls)
